core/runtime.rs

```diff
@@ -180,6 +184,57 @@ pub struct JsRuntimeState {
   waker: AtomicWaker,
 }
 
+impl JsRuntimeState {
+  pub(crate) fn pending_ops(&self, realm_idx: usize) -> &FuturesUnordered<OpCall> {
+    let JsRealmState::Active(_, ops) = self.known_realms.get(realm_idx).unwrap() else {
+      unreachable!()
+    };
+    ops
+  }
+
+  pub(crate) fn pending_ops_mut(&mut self, realm_idx: usize) -> &mut FuturesUnordered<OpCall> {
+    let JsRealmState::Active(_, ops) = self.known_realms.get_mut(realm_idx).unwrap() else {
+      unreachable!()
+    };
+    ops
+  }
+
+  pub(crate) fn context<'s>(&self, realm_idx: usize, scope: &mut v8::HandleScope<'s, ()>) -> Option<v8::Local<'s, v8::Context>> {
+    let JsRealmState::Active(weak, _) = self.known_realms.get(realm_idx).unwrap() else {
+      return None;
+    };
```

Comment:
Why can we return `None` here? Isn't it a coding error if the realm is disposed to call this function?

Reply:
This is where it gets complex. As realms exist right now, they are only ever created and never destroyed.

I think I have to clean this code up quite a bit -- if a realm is disposed, it doesn't actually clean up the associated runtime data, just clears the slots in its context which triggers the GC where we can actually clean things up. This means that there's a period of time where we have a realm that's half disposed.

I'm starting to think we should avoid realm indexes and just use Rc<something> instead. Let me try a different approach as I'm not terribly happy with this.

Reply:
Ah I see, that makes sense. Feel free to address it in a followup if you prefer.